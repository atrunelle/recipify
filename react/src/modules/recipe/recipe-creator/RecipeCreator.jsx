import './RecipeCreator.css';

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Grid } from '@material-ui/core';

import SearchIngredient from '../components/search-ingredient/SearchIngredient';
import IngredientsList from '../components/ingredients-list/IngredientsList';
import IngredientsData from '../components/ingredients-data/IngredientsData';
import * as ingredientsActions from '../store/actions/ingredientsActions';
import * as recipesActions from '../store/actions/recipesActions';
import recipeService from '../recipe.service';
import NutritionMacro from '../components/nutrition-macro/NutritionMacro';
import nutritionService from '../nutrition.service';

class RecipeCreator extends Component {
	addIngredient = (ingredientName, numberOfServings) => {
		this.props.dispatch(ingredientsActions.fetchIngredientPending());

		recipeService.fetchIngredient(ingredientName, numberOfServings)
			.then((data) => {
				this.props.dispatch(ingredientsActions.addIngredient(data));
				this.calculateTotalNutriton();
			})
			.catch((error) => {
				this.props.dispatch(ingredientsActions.fetchIngredientRejected(error));
			});
  }

  removeIngredient = (index) => {
		if (this.props.ingredients.length === 1) {
			this.removeAllIngredients();
		} else {
			this.props.dispatch(ingredientsActions.removeIngredient(index));
			this.calculateTotalNutriton();
		}
  }

  removeAllIngredients = () => {
		this.props.dispatch(ingredientsActions.removeAllIngredients());
		this.props.dispatch(ingredientsActions.resetTotalCalories());
		this.props.dispatch(ingredientsActions.resetTotalWeight());
		this.props.dispatch(ingredientsActions.resetTotalNutritients());
	}
	
	saveRecipe = (name) => {
		const recipe = {
			ingredients: [ ...this.props.ingredients ],
			name,
		};
		this.removeAllIngredients();
		this.props.dispatch(recipesActions.saveRecipe(recipe));
	}

	calculateTotalNutriton = () => {
		const totalCalories = nutritionService.getTotalCalories(this.props.ingredients);
    const totalWeight = nutritionService.getTotalWeight(this.props.ingredients);

		this.props.dispatch(ingredientsActions.updateTotalCalories(totalCalories));
		this.props.dispatch(ingredientsActions.updateTotalWeight(totalWeight));

		const totalNutrients = nutritionService.getTotalNutrients(this.props.ingredients, this.props.totalCalories);
		this.props.dispatch(ingredientsActions.updateTotalNutriton(totalNutrients));
	}

	render() {
		return (
			<Grid container spacing={16}>
				<Grid item xs={12}>
					<SearchIngredient
						add={this.addIngredient}
						isFetching={this.props.isFetching}/>
				</Grid>
				<Grid item xs={12}>
					<IngredientsList 
						ingredients={this.props.ingredients}
						remove={this.removeIngredient}
						removeAll={this.removeAllIngredients}
						save={this.saveRecipe}/>
				</Grid>
				<Grid item xs={12}>
					<IngredientsData 
						totalNutrients={this.props.totalNutrients}
						totalWeight={this.props.totalWeight}
						totalCalories={this.props.totalCalories}/>
				</Grid>
				<Grid item xs={12}>
					<NutritionMacro totalNutrients={this.props.totalNutrients}/>
				</Grid>
			</Grid>
		);
	}
}

const mapStateToProps = (state) => ({
	ingredients: state.ingredients.ingredients,
	totalNutrients: state.ingredients.totalNutrients,
	totalWeight: state.ingredients.totalWeight,
	totalCalories: state.ingredients.totalCalories,
	isFetching: state.ingredients.isFetching,
});

export default connect(mapStateToProps)(RecipeCreator);