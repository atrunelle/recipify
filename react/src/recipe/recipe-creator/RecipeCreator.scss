.search {
    padding: 20px;

    &__input {
        height: 40px;
        width: 100%;
    }

    &__label {
        font-weight: bold;
        margin-bottom: 10px;
        display: block;
    }
}

