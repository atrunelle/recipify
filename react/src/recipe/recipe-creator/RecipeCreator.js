import './RecipeCreator.css';

import React, { Component } from 'react';
import { connect } from 'react-redux';
import apiService from '../../core/api.service';

import SearchIngredient from '../components/search-ingredient/SearchIngredient';
import IngredientsList from '../components/ingredients-list/IngredientsList';
import IngredientsData from '../components/ingredients-data/IngredientsData';
import * as ingredients from '../actions/ingredientsActions';

class RecipeCreator extends Component {

	getIngredients(data) {
		const MEASUREMENT = 'Kilogram';

		const measureUri = data
			.map((d) => d.measures
				.find(m =>  m.label === MEASUREMENT)
			);

		const ingredients = data.map((item, index) => {
			return {
				quantity: 1,
				measureUri: measureUri[index].uri,
				foodURI: item.food.uri,
			}
		});

		return {
			ingredients,			
		}			
	}

	search(value) {
		 apiService.getItem(value).then((data) => {
			const ingredients = this.getIngredients(data.hints);

			apiService.getNutrients(ingredients)
				.then((results) => {
					this.setState({
						value,
						ingredients: results,
					});

				});
		});
  }

  addIngredient = () => {
    this.props.dispatch(ingredients.addIngredient());
  }

  removeIngredient = (index) => {
    this.props.dispatch(ingredients.removeIngredient(index));
  }

  removeAllIngredients = () => {
    this.props.dispatch(ingredients.removeAllIngredients());
  }

	render() {
		return (
			<div>
				<SearchIngredient add={this.addIngredient}/>
				<IngredientsList 
          ingredients={this.props.ingredients}
          remove={this.removeIngredient}
          removeAll={this.removeAllIngredients}/>
        <IngredientsData ingredients={this.props.ingredients}/>
			</div>
		);
	}
}

const mapStateToProps = (state) => ({
	ingredients: state.ingredients,
});

export default connect(mapStateToProps)(RecipeCreator);