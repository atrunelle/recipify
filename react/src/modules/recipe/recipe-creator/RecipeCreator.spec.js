import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import RecipeCreator from './RecipeCreator';

describe('Component: RecipeCreator', () => {
  const initialState = {
    ingredients: [],
    totalNutrients: [],
    totalCalories: 0,
    totalWeight: 0,
  };
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it('should match snapshot', () => {
    const component = mount(<Provider store={store}>
      <RecipeCreator />
    </Provider>);
    expect(component.html()).toMatchSnapshot();
  });
});