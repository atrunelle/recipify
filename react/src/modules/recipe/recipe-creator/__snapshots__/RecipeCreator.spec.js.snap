// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: RecipeCreator should match snapshot 1`] = `"<div class=\\"MuiGrid-container-1 MuiGrid-spacing-xs-16-23\\"><div class=\\"MuiGrid-item-2 MuiGrid-grid-xs-12-39\\"><div class=\\"MuiPaper-root-93 MuiPaper-elevation2-97 MuiPaper-rounded-94 MuiCard-root-92 u-push-top\\"><div class=\\"MuiCardContent-root-120\\"><h2 class=\\"MuiTypography-root-121 MuiTypography-headline-126 MuiTypography-gutterBottom-138\\">What do you have in your fridge?</h2><form class=\\"search\\"><div class=\\"MuiGrid-container-1 MuiGrid-spacing-xs-8-22 MuiGrid-justify-xs-space-between-20\\"><div class=\\"MuiGrid-item-2 MuiGrid-grid-xs-12-39 MuiGrid-grid-sm-6-46\\"><div class=\\"MuiFormControl-root-145 MuiFormControl-fullWidth-148\\"><label class=\\"MuiFormLabel-root-154 MuiInputLabel-root-149 MuiInputLabel-formControl-150 MuiInputLabel-animated-153\\" data-shrink=\\"false\\" for=\\"ingredientNameInput\\">Find an ingredient</label><div class=\\"MuiInput-root-159 MuiInput-fullWidth-166 MuiInput-formControl-160 MuiInput-underline-163\\"><input type=\\"text\\" aria-invalid=\\"false\\" aria-required=\\"false\\" class=\\"MuiInput-input-167\\" id=\\"ingredientNameInput\\" placeholder=\\"Ex: 200gr chicken\\" value=\\"\\"></div></div></div><div class=\\"MuiGrid-item-2\\"><div class=\\"MuiFormControl-root-145\\"><label class=\\"MuiFormLabel-root-154 MuiInputLabel-root-149 MuiInputLabel-formControl-150 MuiInputLabel-animated-153 MuiInputLabel-shrink-152\\" data-shrink=\\"true\\" for=\\"numberOfServingsInput\\">Number of servings</label><div class=\\"MuiInput-root-159 MuiInput-formControl-160 MuiInput-underline-163\\"><input type=\\"text\\" aria-invalid=\\"false\\" aria-required=\\"false\\" class=\\"MuiInput-input-167\\" id=\\"numberOfServingsInput\\" value=\\"1\\"></div></div></div><button tabindex=\\"-1\\" class=\\"MuiButtonBase-root-188 MuiButtonBase-disabled-189 MuiButton-root-172 MuiButton-raised-178 MuiButton-raisedPrimary-179 MuiButton-disabled-182\\" type=\\"submit\\" disabled=\\"\\" id=\\"submitButton\\"><span class=\\"MuiButton-label-173\\">Add ingredient</span></button></div></form></div></div></div><div class=\\"MuiGrid-item-2 MuiGrid-grid-xs-12-39\\"></div><div class=\\"MuiGrid-item-2 MuiGrid-grid-xs-12-39\\"></div><div class=\\"MuiGrid-item-2 MuiGrid-grid-xs-12-39\\"></div></div>"`;
